MDANSE_GUI/Src/MDANSE_GUI/Tabs/Models/TrajectoryModel.py

```diff
@@ -0,0 +1,156 @@
+#    This file is part of MDANSE_GUI.
+#
+#    MDANSE_GUI is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU General Public License as published by
+#    the Free Software Foundation, either version 3 of the License, or
+#    (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+from qtpy.QtCore import (
+    QDeadlineTimer,
+    QModelIndex,
+    QMutex,
+    QObject,
+    Qt,
+    QThread,
+    Signal,
+    Slot,
+)
+from qtpy.QtGui import QStandardItem, QStandardItemModel
+
+from MDANSE.MLogging import LOG
+from MDANSE.MolecularDynamics.Trajectory import Trajectory
+
+
+class LoaderThread(QThread):
+    results = Signal(object)
+
+    def __init__(
+        self,
+        parent,
+        filename: str,
+        index: int,
+        *,
+        display_item: QStandardItem | None = None,
+        target_label: str | None = None,
+    ):
+        super().__init__(parent)
+        self._filename = filename
+        self._target_index = index
+        self._display_item = display_item
+        self._target_label = target_label
+
+    def run(self):
+        trajectory = Trajectory(self._filename)
+        self.results.emit((trajectory, self._target_index))
+        if self._display_item is not None and self._target_label is not None:
+            self._display_item.setData(
+                self._target_label, role=Qt.ItemDataRole.DisplayRole
+            )
+
+
+class TrajectoryModel(QStandardItemModel):
+    """Like GeneralModel, but should implement trajectory
+    loading in the background."""
+
+    error = Signal(str)
+    all_elements = Signal(object)
+    finished_loading = Signal(int)
+    free_name = Signal(str)
+
+    def __init__(self, parent: QObject = None):
+        super().__init__(parent=parent)
+        self.mutex = QMutex()
+        self._node_numbers = []
+        self._trajectory_paths = {}
+        self._trajectory_instances = {}
+        self._loading_threads = {}
+        self._next_number = 0
```

Comment:
Probably want to use `itertools.count` with `next`
```suggestion
        self._next_number = itertools.count()
```